### DISCUSSION TITLE
sim-dev default branch?

### DISCUSSION THREAD

**Discussion by jacueblol** [Patribots4738/Crescendo2024#153] [Q&A]
Sim dev is going to essentially be our playground for any and all new things on the bot; I think that it would be really nice to have the sim-dev as the default branch to make the workflow easier, yet of course, have protections for the main branch so nothing happens to it. What does everyone think?

**Comment by GalexY727** [comment]
it's nice to have an ol' reliable branch of the code, but I don't really see main getting used outside of updating it for competition... Is that a bad thing? I really don't know D:

**Comment by GalexY727** [comment]
We could just make releases of the main branch as the competition checkmarks and we can go back to those releases if there are major issues, OR have a branch for comps which I plan on doing anyway, I just am back to where I was originally about not needing a dev branch since we are going to never _not_ want to use it over the main branch when using the real robot.

**Reply by GalexY727** [reply]
I biggest problem is that we need students to know that they need to test their code in sim before they make a PR to main.

**Comment by jacueblol** [comment]
I love having a dev branch; I think that it is very nice to check code between multiple different issue branches before going to the main.
I think it should stay; however, that is not the discussion I want here. I want this discussion to be about whether the dev should be switched to the default branch, as it would make a new issue automatic on selecting a source. 

on a side note, I think that making versions for comps is a great thing to do, as we can keep track of it a bit better than scrolling through the main.

**Reply by GalexY727** [reply]
if it's the default branch it would be equivalent to removing it and having it as main instead. I think that the default branch should always either be named "main" or "master" to align with the standards for source control across the world.

**Reply by jacueblol** [reply]
I don't think that it would make it irrelevant, but I do think that main should be a default, I wish there was a way that you could make the source be a different branch other than main on a new PR

**Reply by GalexY727** [reply]
if we had two issues, let's call them #998 and #999, and both were forked off of main at the same time: and they both needed to go into main together-- then I think that #999 should merge into #998 and _that_ should be merged into main. If #999 was done and dusted before #998, then #999 should go into main and #998 would later go into main after it itself finishes. It would need to pull main after it gets #999 merged into it and that in itself would prompt more testing to make sure that all of the merge conflicts are resolved and #998 would be tested to see if it still works before going into main.

**Reply by GalexY727** [reply]
> I don't think that it would make it irrelevant, but I do think that main should be a default, I wish there was a way that you could make the source be a different branch other than main on a new PR

I see this making main get lost in the dark, never to be edited again...

**Reply by jacueblol** [reply]
> if we had two issues, let's call them #998 and #999, and both were forked off of main at the same time: and they both needed to go into main together-- then I think that #999 should merge into #998 and _that_ should be merged into main. If #999 was done and dusted before #998, then #999 should go into main and #998 would later go into main after it itself finishes. It would need to pull main after it gets #999 merged into it and that in itself would prompt more testing to make sure that all of the merge conflicts are resolved and #998 would be tested to see if it still works before going into main.

I see what u mean, but that defeats the purpose of main, which is a branch that you can say for certain that it is ALWAYS going to work on the bot.

**Reply by jacueblol** [reply]
> > I don't think that it would make it irrelevant, but I do think that main should be a default, I wish there was a way that you could make the source be a different branch other than main on a new PR
> 
> I see this making main get lost in the dark, never to be edited again...

Going back to my comment above, making dev the source for all PRs would not make main irrelivent. dev should be a source for new ideas, while main is a means for deployment. If dev was the source for all PRs we could be sure that the commits would always have to be tested in sim before reaching the bot. again making main the source for the bot's working code, not the code in developement.

**Reply by GalexY727** [reply]
it's hard for me to imagine a change that would break the bot that would be in sim-dev... are you considering it as a branch that should be used as little as possible so that we have an incentive to test it and get it to main fast

**Reply by jacueblol** [reply]
> it's hard for me to imagine a change that would break the bot that would be in sim-dev... are you considering it as a branch that should be used as little as possible so that we have an incentive to test it and get it to main fast

yes, my vision for the workflow is this:
PR is made for development off of sim-dev ( by default, but if required could be a different branch ) >> changes are made >> PRed back to dev ( most likely it will be working in tandom with another issue or by default needs to be tested in sim ) >> PRed to main to update main code, with everything running smooth as butter on main. 

if something doesn't work then that means that its just a small bot thing that doesn't translate immediately across to sim. we've gotten this far this fast  because we have had sim and I think that is an important thing to highlight

**Comment by jacueblol** [comment]
the final thing we landed on is this kinda workflow, but only at the start of the season. It gets hard to work with later in the season.
